cassiopeia/data/CassiopeiaTree.py

```diff
@@ -68,117 +73,291 @@ class CassiopeiaTree:
 
     def __init__(
         self,
-        character_matrix: pd.DataFrame,
+        character_matrix: Optional[pd.DataFrame] = None,
+        missing_state_indicator: int = -1,
         cell_meta: Optional[pd.DataFrame] = None,
         character_meta: Optional[pd.DataFrame] = None,
         priors: Optional[Dict[int, Dict[int, float]]] = None,
-        tree: Optional[Union[str, ete3.Tree, nx.DiGraph]] = None
+        tree: Optional[Union[str, ete3.Tree, nx.DiGraph]] = None,
     ):
 
-        self.character_matrix = character_matrix
+        self.__original_character_matrix = None
+        self.__current_character_matrix = None
+        if character_matrix is not None:
+            self.__original_character_matrix = character_matrix.copy()
+            self.__current_character_matrix = character_matrix.copy()
+        
+        self.missing_state_indicator = missing_state_indicator
         self.cell_meta = cell_meta
         self.character_meta = character_meta
         self.priors = priors
         self.__network = None
 
-        self.populate_tree(tree)
+        if tree is not None:
+            self.populate_tree(tree)
 
     def populate_tree(self, tree: Union[str, ete3.Tree, nx.DiGraph]):
 
         if isinstance(tree, nx.DiGraph):
             self.__network = tree
-        if isinstance(tree, str):
+        elif isinstance(tree, str):
             self.__network = utilities.newick_to_networkx(tree)
-        if isinstance(tree, ete3.Tree):
+        elif isinstance(tree, ete3.Tree):
             self.__network = utilities.ete3_to_networkx(tree)
+        else:
+            raise CassiopeiaTreeError(
+                "Please pass an ete3 Tree, a newick string, or a Networkx object."
+            )
+
+        # add character states
+        for n in self.nodes:
+            if (
+                self.__original_character_matrix is not None
+                and n in self.__original_character_matrix.index.values
+            ):
+                self.__network.nodes[n][
+                    "character_states"
+                ] = self.__original_character_matrix.loc[n].to_list()
+            else:
+                self.__network.nodes[n]["character_states"] = []
+
+        # instantiate branch lengths
+        for u, v in self.edges:
+            self.__network[u][v]["length"] = 1
+
+        # instantiate node time
+        self.__network.nodes[self.root]["time"] = 0
+        for u, v in self.depth_first_traverse_edges(source=self.root):
+            self.__network.nodes[v]["time"] = (
+                self.__network.nodes[u]["time"] + self.__network[u][v]["length"]
+            )
+
+    def initialize_character_states_at_leaves(
```

Comment:
I think we should make this naming convention more clear, like "initialize_character_matrix_from_leaves"

Reply:
I found the name `initialize_character_states_at_leaves` was already clear, in fact I find the alternative name `initialize_character_matrix_from_leaves` a bit confusing? The `character_matrix` of the CassiopeiaTree is being initialized from the `character_matrix` passed as argument, not from "leaves"? (What exactly do you mean by "leaves?") Or did you mean `initialize_character_matrix_at_leaves`? Or maybe just `initialize_character_matrix`?